text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,237 @@
+  I believe this RFC satisfies the constraints outlined, mostly by dropping `DynSized` as an available bound, which limits expressiveness in favour of simplicity.
+- [Custom DSTs](https://github.com/rust-lang/rfcs/pull/2594).  
+  This postponed RFC attempts to introduce a generic framework for DSTs with arbitrary metadata.
+  This RFC aims to be compatible with future attempts at custom DSTs, as it is in essence a very restricted form.
+- [DynSized without ?DynSized](https://github.com/rust-lang/rfcs/pull/2310).  
+  This contains a lot of very useful analysis of what exactly commonly used crates want when they state `?Sized`.
+  However, this RFC aims to be simpler than the lint-based solution presented there.
+  Additionally, this deals with not being able to place extern types as fields in structs, rather than solely on `size_of_val` and `align_of_val`.
+- [More implicit bounds](https://github.com/rust-lang/rfcs/issues/2255).  
+  This discusses whether we should be adding more implicit bounds into the language, and there associated relaxations (like `?MetaSized`).
+  This RFC aims to sidestep this issue by utilising the edition system to change the definition of `?Sized`.
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should `MetaSized` contain methods, or should it be a marker trait with the implementation left to `core::mem::size_of_val` and `core::mem::align_of_val`?
+- Should "metadata sized" imply "metadata aligned" or should we be adding the `MetaAligned` trait rather than `MetaSized`?
+- Should `MetaSized` be a supertrait of `Sized`? All `Sized` things are `MetaSized` but `Sized` doesn't semantically require `MetaSized`.
+- Should users be able to slap a `#[repr(align(n))]` attribute onto opaque types to give them an alignment?  
+  This would allow us to represent `CStr` properly but would necessitate splitting `MetaSized` and `MetaAligned` as they would not be "metadata sized" in general.
+  (We may be able to get away with the [Aligned trait](https://github.com/rust-lang/rfcs/pull/3319))
+- Should the `extern type` syntax exist, or should there just be a `repr(unsized)`?  
+  This would allow headers with opaque tails (which are very common in C code) but is a more significant departure from the original RFC, and looks more like custom DSTs.
+
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+The most obvious future possibility is custom DSTs.
+This would provide a mechanism for allowing users to implement types with entirely custom metadata and therefore custom implementations of `size_of_val` and `align_of_val`.
+This would likely mean that users could implement types that acted like extern types without using the `extern type` syntax.
+This should not be an issue as `extern type` communicates the intent of these types well, and guarantees FFI-safety.
+
```

Comment:
One other valuable extension would be to make `MaybeUninit` accept `?Sized + MetaSized` types.

Reply:
MaybeUninit currently takes a sized T, why would this enable relaxing that bound? `T: ?Sized` currently means the same as I'm proposing `T: ?Sized + MetaSized` will mean, so I don't see how this helps.

Reply:
Oh, I misread the proposal and assumed that `?Sized` would opt out of `MetaSized` too.

Without a notion of `MetaSized`, `MaybeUninit` has to be sized, since there's no good way to determine the layout of an unsized value. Adding `MetaSized` means we can now do this with just metadata, making it valid again.

Reply:
Sorry to be clear the proposal is that `?Sized` will opt out of `MetaSized` after the edition change. So far I've assumed that, because all types in current Rust are `MetaSized`, that introducing this won't help with issues like this. It's possible that's not true and this is an exception.

How would you create a `MaybeUninit<T: ?Sized + MetaSized>`, where would you get the metadata from? Can it be uninit? Aren't those the same issues you'd face if you opened an RFC to make `MaybeUninit` take `?Sized` today?